lkcrypto.c

```diff
@@ -209,7 +276,104 @@ static int decrypt_encrypt(lua_State *L, int dec_enc)
 	} else {
 		kfree(buffer);
 		return luaL_error(L, "%s failed %d",
-				  dec_enc ? "encrypt" : "decrypt");
+				  operation ? "encrypt" : "decrypt");
+	}
+}
+
+static int decrypt_encrypt_aead(lua_State *L, enum ENC_DEC operation)
+{
+	size_t key_len = 0;
+	size_t iv_len = 0;
+	size_t auth_len = 0;
+	size_t ciphertext_len = 0;
+	size_t aad_len = 0;
+
+	int result = 0;
+	int nargs = 0;
+	const u8 *key;
+	const u8 *iv_orig;
+	const u8 *auth_orig;
+	const u8 *aad_orig;
+	const u8 *ciphertext;
+	u8 iv[AES_MAX_KEY_SIZE] = {
+		0,
+	};
+	u8 *auth = NULL;
+	u8 *buffer = NULL;
+	u8 *aad = NULL;
+
+	struct cipher_aead_tfm_t *tfm = NULL;
+
+	nargs = lua_gettop(L);
+
+	tfm = luaL_checkudata(L, 1, CUJO_CIPHER_AEAD_TFM_LUA_NAME);
+	key = luaL_checklstring(L, 2, &key_len);
+	ciphertext = luaL_checklstring(L, 3, &ciphertext_len);
+	iv_orig = luaL_checklstring(L, 4, &iv_len);
+	memcpy(iv, iv_orig, iv_len);
+	if (operation == DECODING) {
+		auth_orig = luaL_checklstring(L, 5, &auth_len);
+		auth = kmalloc(auth_len, GFP_ATOMIC);
+		if (auth == NULL) {
+			return luaL_error(L, "auth allocation failed");
+		}
+		memcpy(auth, auth_orig, auth_len);
+	}
+
+	buffer = kmalloc(ciphertext_len, GFP_ATOMIC);
+	if (buffer == NULL) {
+		return luaL_error(L, "buffer allocation failed");
+	}
+
+	if (operation == ENCODING) {
+		auth_len = key_len;
+		auth = kmalloc(auth_len, GFP_ATOMIC);
+		if (auth == NULL) {
+			return luaL_error(L, "auth allocation failed");
+		}
+		if (nargs == 5) {
+			aad_orig = luaL_checklstring(L, 5, &aad_len);
+			aad = kmalloc(aad_len, GFP_ATOMIC);
+			if (aad == NULL) {
+				return luaL_error(L, "aad allocation failed");
```

Comment:
Missing `kfree(buffer)` in both of these returns.

There are enough error paths here that you may want to consider using `goto` to avoid having to spell out the necessary frees every time.